### DISCUSSION TITLE
GraphQL Query to get user's org memberships with role?

### DISCUSSION THREAD

**Discussion by grahamb** [community/community#24656] [Apps, API and Webhooks]
<p>We have an internal tool for helping manage our GHES organizations. I’m trying to find the most efficient way of obtaining a list of organizations for a given user, with their role in each organization (<code>MEMBER</code> or <code>ADMIN</code>). The goal is to be able to show the user a list of all their org memberships with their role; a user with an <code>ADMIN</code> org role would then be able to have access to some management tools.</p>
<p>I’m making these queries as an admin user who is a member of all our orgs. It seems like the only way to do this is:</p>
<p><strong>Option 1</strong>: Make a <code>user</code> query including the <code>organizations</code> field, which may be paginated depending on how many orgs the <code>user</code> is a member of. If <code>organizations</code> is paginated, we need to follow that to resolve all their org memberships. The <code>membersWithRole</code> field is also paginated, so that also needs to be followed until there’s no further records.</p>
<p>Once we have all the <code>membersWithRole</code> for each <code>orgainzation</code>, we have to then find the <code>membersWithRole</code> node that matches the user’s <code>login</code> and see what <code>role</code> they have.</p>
<pre><code class="lang-auto">query getUserWithOrganizations($login: String!) {
  user(login: $login) {
    login
    organizations(first: 100) { # pagination needs to be resolved!
      totalCount
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
      edges {
        node {
          id
          login
          name
          description
          membersWithRole(first: 100) { # more pagination!
            pageInfo {
              hasNextPage
              hasPreviousPage
              startCursor
              endCursor
            }
            edges {
              role
              node {
                login
              }
            }
          }
        }
      }
    }
  }
}
</code></pre>
<p><strong>Option 2</strong>: Make a <code>user</code> query including the <code>organizations</code> field, but only for <code>organizatins.login</code>. Resolve that pagination tree. When we have a list of all the user’s orgs, make individual queries for each <code>organization</code> with <code>membersWithRole</code>, resolving that pagination, and suss out what role the user has in each org.</p>
<pre><code class="lang-auto">
# make a query to get the user's orgs

query getUserWithOrganizations($login: String!) {
  user(login: $login) {
    login
    organizations(first: 100) {
      totalCount
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
      }
      edges {
        node {
          id
          login
          name
          description
        }
      }
    }
  }
}

# then, once we have fetched all the orgs, 
# make individual calls for org details including 
# `membersWithRoles`. Once that pagination is resolved, 
# filter the `membersWithRoles` for the user in question 
# to obtain their role in each org

query organizationWithMembers($orgLogin: String!) {
  organization(login: $orgLogin) {
    createdAt
    id
    login
    membersWithRole(first: 100) {
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
      }
      edges {
        role
        cursor
        node {
          login
        }
      }
    }
  }
}
</code></pre>
<p>Both of these approaches seem highly inefficient. Ideally, <code>membersWithRole</code> would accept either or both of a <code>role</code> and <code>login</code> filter but that isn’t the case.</p>
<p>Is there a better way to do what I’m trying to accomplish here? I’m not limited to GraphQL, REST is fine, but I can’t see a way to do this with the REST API either.</p>
<p>I do have an alternative way of getting the ownership info; we manage our org membership via our in-house group management tools, and we have an external mapping of org → member_groups and org → admin_groups, so I can always use the user’s membership info from our groups, but I think that there should be a better way of getting this kind of information from the APIs.</p>

**Comment by jsoref** [comment]
<p>I’ve run into a number of limitations like this with the GitHub API. Thankfully in my case when I needed permissions information, there was a local bit, so I could cheat:</p><aside class="onebox githubblob" data-onebox-src="https://github.com/check-spelling/check-spelling/blob/cb7b743c240a74786679a4de908dbe5b013984ce/unknown-words.sh#L331">
  <header class="source">

      <a href="https://github.com/check-spelling/check-spelling/blob/cb7b743c240a74786679a4de908dbe5b013984ce/unknown-words.sh#L331" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/check-spelling/check-spelling/blob/cb7b743c240a74786679a4de908dbe5b013984ce/unknown-words.sh#L331" target="_blank" rel="noopener nofollow ugc">check-spelling/check-spelling/blob/cb7b743c240a74786679a4de908dbe5b013984ce/unknown-words.sh#L331</a></h4>



    <pre class="onebox">      <code class="lang-sh">
        <ol class="start lines" start="321" style="counter-reset: li-counter 320 ;">
            <li>pull_request_url=$(jq -r '.pull_request.url // empty' $issue)</li>
            <li>pull_request_info=$(mktemp_json)</li>
            <li>pull_request "$pull_request_url" | jq .head &gt; $pull_request_info</li>
            <li>pull_request_sha=$(jq -r '.sha // empty' $pull_request_info)</li>
            <li>set_comments_url "$GITHUB_EVENT_NAME" "$GITHUB_EVENT_PATH" "$pull_request_sha"</li>
            <li>react_prefix_base="Could not perform [request]($trigger_comment_url)."</li>
            <li>react_prefix="$react_prefix_base"</li>
            <li>if [ "$sender_login" != "$issue_user_login" ]; then</li>
            <li>  collaborators_url=$(jq -r '.repository.collaborators_url // empty' "$GITHUB_EVENT_PATH")</li>
            <li>  collaborators_url=$(echo "$collaborators_url" | perl -pne "s&lt;\{/collaborator\}&gt;&lt;/$sender_login/permission&gt;")</li>
            <li class="selected">  collaborator_permission=$(collaborator "$collaborators_url" | jq -r '.permission // empty')</li>
            <li>  case $collaborator_permission in</li>
            <li>    admin)</li>
            <li>      ;;</li>
            <li>    write)</li>
            <li>      ;;</li>
            <li>    *)</li>
            <li>      confused_comment "$trigger_comment_url" "Commenter (@$sender_login) isn't author (@$issue_user_login) / collaborator"</li>
            <li>      ;;</li>
            <li>  esac</li>
            <li>fi</li>
        </ol>
      </code>
    </pre>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>I certainly haven’t seen any <code>joins</code> for things like this. And the number of queries I’d need to do certainly worries me (both in terms of query time and quota count).</p>
<p>Personally, if I had access to a private resource with the information I needed, I’d use it.</p>
<p>The closest I have to your case is here:</p><aside class="onebox githubblob" data-onebox-src="https://github.com/check-spelling/check-spelling/blob/e2ab5ecbd692e97052084ea6a8fc14ced3784b67/check-pull-requests.sh#L34">
  <header class="source">

      <a href="https://github.com/check-spelling/check-spelling/blob/e2ab5ecbd692e97052084ea6a8fc14ced3784b67/check-pull-requests.sh#L34" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/check-spelling/check-spelling/blob/e2ab5ecbd692e97052084ea6a8fc14ced3784b67/check-pull-requests.sh#L34" target="_blank" rel="noopener nofollow ugc">check-spelling/check-spelling/blob/e2ab5ecbd692e97052084ea6a8fc14ced3784b67/check-pull-requests.sh#L34</a></h4>



    <pre class="onebox">      <code class="lang-sh">
        <ol class="start lines" start="24" style="counter-reset: li-counter 23 ;">
            <li>tree_config="$bucket/$project/"</li>
            <li>stored_config=$temp/config/</li>
            <li>headers=$temp/headers</li>
            <li>
            </li>
<li>Q='"'</li>
            <li>
            </li>
<li>owner=${GITHUB_REPOSITORY%/*}</li>
            <li>repo=${GITHUB_REPOSITORY#*/}</li>
            <li>length=20</li>
            <li>
            </li>
<li class="selected">get_open_pulls() {</li>
            <li>query=$(echo "query {</li>
            <li>  repository(owner:${Q}$owner${Q} name:${Q}$repo${Q}) {</li>
            <li>    pullRequests(first: $length states: OPEN $continue) {</li>
            <li>      totalCount</li>
            <li>      pageInfo {</li>
            <li>        hasNextPage</li>
            <li>        endCursor</li>
            <li>      }</li>
            <li>      nodes {</li>
            <li>        baseRefName</li>
        </ol>
      </code>
    </pre>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>And I punted. In theory a branch could be tied to a PR in any repository in its fork network. I decided that running a paginated query to get all repositories in the fork network (possibly greater than 1000, i.e. greater than 50 pages) and then performing pagination for my per repository (individually for each of the 1000 repositories) query was just going to be too expensive.</p>
<p>In fact, these sorts of painful edges are why I joined github.community.</p>

**Comment by byrneh** [comment]
<p>Hi <a class="mention" href="/grahamb">@grahamb</a> I had a look for you also but no joy.</p>
<p>Nested paging is a pain, complexity and many not be efficient either :frowning:<br>
If you know (in your specific circumstance) you will not have more that 100 organizations or more than 100 membersWithRole within a single organization you may be able to ignore paging for that edge that returns less than 100 items.<br>
if you are handling pagination you will need to use<br>
<code>( first: 100, after: $endCursor )</code></p>
<p><code>On the membersWithRole</code> ( <code>OrganizationMemberConnection!</code> ) there is is <strong>role</strong> field on the edge to retrieve the role MEMEBR or ADMIN value of each member listed</p>
<pre><code class="lang-auto">membersWithRole ( first: $first, after: $endCursor ) {
	totalCount
	edges {
		role
		node { login }
	}
	pageInfo { endCursor, hasNextPage }
}
</code></pre>
<p>The bad news is. I also could not find a away to limit a query at the organization level to just a specific login (user) and the membership role and hence be  ore efficient:(<br>
The documentation states membersWithRole only accepts arguments <code>(after, before, first, last)</code>, which as you say for your requirement we ned to be able to filter on <strong>login for the user being queried</strong> but cannot.</p>

**Comment by byrneh** [comment, marked as answer]
<p>Hi <a class="mention" href="/grahamb">@grahamb</a><br>
There is a REST API</p>
<blockquote>
<p>GET /orgs/{org}/memberships/{username}</p>
</blockquote>
<p>and the json retuned is just for that username and includes <strong>role</strong> field on the organization for the username.</p>
<p>So this is an option that could be used in conjunction with a retrieved list of  user organizations</p>

**Comment by grahamb** [comment]
<p><a class="mention" href="/byrneh">@byrneh</a> Unfortunately we do have at least one org with over 100 members so ignoring gql pagination won’t work, but for my needs that REST endpoint will certainly do the trick. Thanks for the tip!</p>
